File: backend/app/api/claude.py
from fastapi import APIRouter, Depends

from app.api.deps import verify_token
from app.db import jobs_repo
from app.schemas.claude import ClaudeSessionResponse, ClaudeSpawnRequest
from app.schemas.jobs import JobResponse
from app.services.jobs import enqueue_job
from app.services.sessions import create_session
from app.websocket.manager import manager

router = APIRouter(prefix="/claude")


@router.post("/spawn", response_model=JobResponse)
async def spawn_claude_api(
    request: ClaudeSpawnRequest, _: None = Depends(verify_token)
) -> JobResponse:
    payload = request.model_dump()
    job_id = await jobs_repo.create_job("claude.spawn", payload)
    await enqueue_job(job_id)
    await manager.emit_log("info", f"claude job queued {job_id}")
    return JobResponse(job_id=job_id, status="queued")


@router.post("/session", response_model=ClaudeSessionResponse)
async def create_claude_session(
    _: None = Depends(verify_token),
) -> ClaudeSessionResponse:
    session_id = await create_session()
    await manager.emit_log("info", f"claude session created {session_id}")
    return ClaudeSessionResponse(session_id=session_id)

File: backend-bun/src/routes/claude.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { authMiddleware, requireAuth } from "../middleware/auth.ts";
import * as jobsRepo from "../db/jobs.ts";
import { enqueueJob } from "../services/jobs.ts";
import { createSession } from "../services/sessions.ts";
import { manager } from "../websocket/manager.ts";
import { ClaudeSpawnRequestSchema, type ClaudeSessionResponse } from "../schemas/claude.ts";
import type { JobResponse } from "../schemas/jobs.ts";
import type { Env } from "../lib/types.ts";

export const claudeRoutes = new Hono<Env>();

claudeRoutes.use("*", authMiddleware);
claudeRoutes.use("*", requireAuth);

claudeRoutes.post("/spawn", zValidator("json", ClaudeSpawnRequestSchema), (c) => {
  const body = c.req.valid("json");
  const payload = {
    args: body.args ?? [],
    prompt: body.prompt,
    stdin: body.stdin,
    session_id: body.session_id,
    timeout_ms: body.timeout_ms,
  };
  const jobId = jobsRepo.createJob("claude.spawn", payload);
  enqueueJob(jobId);
  manager.emitLog("info", `claude job queued ${jobId}`);
  return c.json<JobResponse>({ job_id: jobId, status: "queued" });
});

claudeRoutes.post("/session", (c) => {
  const sessionId = createSession();
  manager.emitLog("info", `claude session created ${sessionId}`);
  return c.json<ClaudeSessionResponse>({ session_id: sessionId });
});
